File: src/main/dart/g0101_0200/s0152_maximum_product_subarray/Solution.dart
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Dynamic_Programming
// #Dynamic_Programming_I_Day_6 #Level_2_Day_13_Dynamic_Programming #Udemy_Dynamic_Programming
// #Big_O_Time_O(N)_Space_O(1) #2024_10_07_Time_347_ms_(81.82%)_Space_149.9_MB_(63.64%)

import 'dart:math';

class Solution {
  int maxProduct(List<int> nums) {
    /*
      m = max 
      l = left 
      r = right
      s = size
      */
    var m = nums.first;
    var l = 1;
    var r = 1;
    var s = nums.length;
    var last = s - 1;
    for (var i = 0; i < s; i++) {
      l *= nums[i];
      r *= nums[last - i];
      m = max(m, l);
      m = max(m, r);
      l = l == 0 ? 1 : l;
      r = r == 0 ? 1 : r;
    }
    return m;
  }
}

File: src/main/js/g0101_0200/s0152_maximum_product_subarray/solution.js
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Dynamic_Programming
// #Dynamic_Programming_I_Day_6 #Level_2_Day_13_Dynamic_Programming #Udemy_Dynamic_Programming
// #Big_O_Time_O(N)_Space_O(1) #2024_12_16_Time_0_ms_(100.00%)_Space_50.9_MB_(51.22%)

/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let overAllMaxProd = Number.MIN_SAFE_INTEGER
    let n = nums.length
    let start = 1
    let end = 1

    for (let i = 0; i < n; i++) {
        // Reset `start` and `end` to 1 if they become 0
        if (start === 0) {
            start = 1
        }
        if (end === 0) {
            end = 1
        }

        start *= nums[i]
        end *= nums[n - i - 1]

        overAllMaxProd = Math.max(overAllMaxProd, Math.max(start, end))
    }

    return overAllMaxProd
};

export { maxProduct }
